### PR TITLE
SQCORE 1173: Servantify the Galley Conversation API

### PR DESCRIPTION
**Pull request by mdimjasevic**
This implements https://wearezeta.atlassian.net/browse/SQCORE-1173. Endpoints:

1) [X] POST /conversations/:cnv/join
2) [X] POST /conversations/code-check
3) [X] POST /conversations/join
4) [X] POST /conversations/:cnv/code
5) [x] DELETE /conversations/:cnv/code
6) [x] GET /conversations/:cnv/code
7) [x] POST /conversations/:cnv/typing
8) [x] POST /broadcast/otr/messages
9) [x] POST /broadcast/otr/messages
10) [x] POST /bot/messages

Note that the last endpoint should technically not be part of this batch, but it makes sense to servantify it together with the other messaging endpoints.

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs [hunk 1]
```diff
@@ -689,17 +814,37 @@ type MessagingAPI =
...
+           ( Summary "Broadcast an encrypted message to all team members and all contacts (accepts JSON or Protobuf)"
+               :> Description PostOtrDescriptionUnqualified
+               :> ZLocalUser
+               :> ZConn
+               :> CanThrow TeamNotFound
+               :> CanThrow BroadcastLimitExceeded
+               :> "broadcast"
+               :> "otr"
+               :> "messages"
+               :> QueryParam "ignore_missing" IgnoreMissing
```
Comment (supersven) [inline libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:839]: Could it be that the Swagger docs of `ignore_missing` and `report_missing` are lost?

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs [hunk 2]
```diff
@@ -689,17 +814,37 @@ type MessagingAPI =
...
+        :> ReqBody '[JSON, Proto] NewOtrMessage
         :> MultiVerb
              'POST
              '[Servant.JSON]
              (PostOtrResponses ClientMismatch)
-             (Either (MessageNotSent ClientMismatch) ClientMismatch)
+             (PostOtrResponse ClientMismatch)
     )
+    :<|> Named
+           "post-otr-broadcast-unqualified"
```
Comment (supersven) [inline libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:829]: Is the former `Error.nonBindingTeam` missing by intention?

**Review by supersven — CHANGES_REQUESTED** [review]
Excellent :+1: 

Some minor things, mostly questions...

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs [hunk 3]
```diff
@@ -831,6 +993,9 @@ type AllFeatureConfigsGet =
         :> Get '[Servant.JSON] AllFeatureConfigs
     )
 
+-- This is a work-around for the fact that we sometimes want to send larger lists of user ids
+-- in the filter query than fits the url length limit.  for details, see
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:997]: Thanks! Will add this in an upcoming push.

**Comment by mdimjasevic** [comment]
@supersven , my apologies, I just realized that I've been rebasing in the middle of reviewing, which is something I suggested you not to do! My bad! I will try not to forget my own suggestions and hopefully I will avoid it in the future.

File: libs/wire-api/src/Wire/API/Routes/Public.hs [hunk 4]
```diff
@@ -153,6 +162,8 @@ type ZConn = ZAuthServant 'ZAuthConn InternalAuthDefOpts
 
 type ZBot = ZAuthServant 'ZAuthBot InternalAuthDefOpts
 
+type ZConversation = ZAuthServant 'ZAuthConv InternalAuthDefOpts
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public.hs:165]: Thanks! Just added it, will be visible in an upcoming commit.

**Comment by supersven** [comment]
> @supersven , my apologies, I just realized that I've been rebasing in the middle of reviewing, which is something I suggested you not to do! My bad! I will try not to forget my own suggestions and hopefully I will avoid it in the future.

No worries. Looks like the comments are still valid and in place. As it's only a handful of comments, I'm pretty sure we can deal with what-ever-might-happen in any case... :smile_cat:

File: services/galley/src/Galley/API/Teams/Common.hs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Galley.API.Teams.Common (getBindingTeam) where
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply services/galley/src/Galley/API/Teams/Common.hs:18]: Good points! I've implemented them in an upcoming commit 7104b1486.

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs (hunk 1 above)
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:839]: They are already contained in the description of the endpoint on line #831: https://github.com/wireapp/wire-server/pull/2016/files#diff-006eafb358b9b080c8fb043d4a24e4575936227b05ae3c8de010af77e1be0febR831.

Comment (supersven) [reply libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:839]: Ah, I see :+1:

File: libs/wire-api/src/Wire/API/Routes/Public/Galley.hs (hunk 2 above)
Comment (mdimjasevic) [reply libs/wire-api/src/Wire/API/Routes/Public/Galley.hs:829]: The error is still here, but now, unlike before, it is tracked via types, hence not manually written via `CanThrow`. If you check the signature of `Galley.API.Update.postOtrBroadcastUnqualified`, which is the handler for this endpoint, you'll see that it has `Error TeamError` as an effect constraint. The team store action combinator you brought up earlier (now called `lookupBindingTeam`) throws a `NotABindingTeamMember`, which in `Wire.API.Error` gets mapped to `nonBindingTeam`:

```haskell
instance APIError TeamError where                                                                                 
...
  toWai NotABindingTeamMember = nonBindingTeam
...
```

**Comment by mdimjasevic** [comment]
@supersven , I believe I addressed all of your comments. Any further comments?

**Review by supersven — APPROVED** [review]
Looks good to me. :+1:

**Comment by mdimjasevic** [comment]
@supersven , sorry, the not-a-binding-team still wasn't showing up! So I added it by hand in a commit e81d8d4b3.

**Pull request merged**